Draft/pythran 0.9.7 (#22)
serge-sans-paille/pythran-stories: content/2020-09-23-pythran-0.9.7.rst

*gast* has moved to **0.4.0** and *beniget* to **0.3.0**, both support Python 3.9
and all these packages are now compatible. Good.

With version 0.9.6, *pythran* introduced a new *numpy* expression computation engine
that solved a few issues but also introduced a performance regression for
various kernels. I've been working on fixing that aspect, and I'm quite happy
with the result, showcased in version 0.9.7.

Performance is a critical aspect of *pythran*, so it comes at no surprise that the
expression evaluation engine got rewritten several times. To evaluate the
difference between 0.9.5 and 0.9.7, let's use the `numpy benchmark
<https://github.com/serge-sans-paille/numpy-benchmarks/>`_ project. It contains
a collection of high-level kernels, and was recently granted a few options to
ease comparison of performance across project version.

.. code-block:: console

    $ pip install pythran==0.9.5
    $ np-bench run -tpythran -p0.9.5- -o 095.log
    $ pip install pythran==0.9.7
    $ np-bench run -tpythran -p0.9.7- -o 097.log
    $ np-bench format 095.log 097.log -tsvg --logscale --normalize=0.9.5-pythran

Which gives:

> .. image:: ./images/2020-09-23-pythran-evolve.svg

Reviewer: Red and green on the same plot makes it **very** difficult for a small but significant fraction of people to really make any sense of the plot (red-green weakness/blindness). Consider C0 and C1? from the mpl palette?

Reviewer 2: [Qualitative colormaps](https://matplotlib.org/tutorials/colors/colormaps.html) should do it. Or: https://seaborn.pydata.org/tutorial/color_palettes.html

> There's quite a few things to tell on that comparison: some benchmarks are in

Reviewer 2: ```suggestion
There are quite a few things to tell on that comparison: some benchmarks are in
```

much better shape (especially ``laplacien``, ``wave`` and ``diffusuon``) but
there's still room for improvement, as shown by ``grayscott`` and
``local_maxima``. The performance boost is due to the better expression engine,

> so that's expected, but the slowdown still need some investigation…

Reviewer 2: ```suggestion
so that's expected, but the slowdown still needs some investigation…
```

The ``np-bench`` script also makes it possible to compare *pythran* with *cpython*
or *numba*. Let's try that:

.. code-block:: console

    $ pip install -U pythran numba
    $ np-bench run -tpythran -tnumba -tpython -oall.log
    $ np-bench format all.log -tsvg --logscale --normalize=python

Which gives:

> .. image:: ./images/2020-09-23-pythran-all.svg

Reviewer: Same as above. Consider C0/1/2? Maybe change linestyles on this image so it works better with greyscale? I'm not sure what bar styles MPL has.

Author: Updated the svg and numpy-benchmark generator accordingly :-)

> Interestingly, unoptimized python is still ahead for a few benchmarks, that
> prove to be too complex to optimize (?) Another subject that needs
> investigation.

Reviewer 2: ```suggestion
Interestingly, unoptimized python is still ahead for a few benchmarks, which
may hint these to be too complex to optimize (?) - another subject that needs
investigation.
```

The kernel are mostly high-level ones, and that doesn't always match *numba*'s

> requirements, which explains that it sometimes just scoops.

Reviewer 2: scoops or stoops?

> Overall *pythran* performance are still satisfying, but we definitively need to

Reviewer 2: ```suggestion
Overall *pythran* performance is still satisfying, but we definitively need to
```
